### PR TITLE
Created posts/saved posts

### PR DESCRIPTION
**Pull request by hewonjeong**
- add createdPost/add
- add savedPosts/add
- add createdPosts/get
- add savedposts/get
- add SavedPosts/delete
- update controller for createdPosts/savedPosts
- update mock data

### CODE REVIEW

**Comment by hewonjeong** [comment]
[[2] As a user, I'd like to see my activity so that I can enjoy my journey in Pingo](https://trello.com/c/owoHXxh7/149-2-as-a-user-i-d-like-to-see-my-activity-so-that-i-can-enjoy-my-journey-in-pingo)

**Comment by hewonjeong** [comment]
[[1] As a user, I'd like to save and see posts in 'favorite' tab so that I can gather interesting posts](https://trello.com/c/WNMwBMTM/150-1-as-a-user-i-d-like-to-save-and-see-posts-in-favorite-tab-so-that-i-can-gather-interesting-posts)

File: src/server/images/models.js [hunk 1]
```diff
@@ -23,4 +25,38 @@ export default class ImageManager {
       cb(new Error('database error'));
     });
   }
+  static getImageUrls(params, cb) {
```
Comment (hewonjeong) [inline src/server/images/models.js]: use literal parameter for readability.
e.g.)getImageUrls({itemKey, userKey}, cb)

File: src/server/users/controllers.js [hunk 2]
```diff
@@ -71,10 +71,9 @@ export default {
       return cb(new APIError(err, { statusCode: err.statusCode, message: err.message }));
     });
   },
-  addCreatedPost(req, res, cb) {
+  addSavedPost(req, res, cb) {
     const timeHash = KeyUtils.genTimeHash();
-    CreatedPostManager.addCreatedPost(req.body.entity, req.body.entityKey, req.body.userKey,
-    timeHash, (err, idxKey) => {
+    SavedPostManager.addSavedPost(req.body.entityKey, req.body.userKey, timeHash, (err, idxKey) => {
```
Comment (hewonjeong) [inline src/server/users/controllers.js]: entityKey?

File: src/server/users/models.js [hunk 3]
```diff
@@ -30,42 +33,194 @@ export default class UserManager {
...
-    const idxPost = {
-      key: entityKey
-    };
-    return db.put(idxKey, idxPost, (err) => {
+    /*
+    make two indexing keys for saved post.
+    1. indexing user : savedPost-{STATE}-{USER_KEY}-{TIMEHASH}
+    2. indexing item : savedPost-{STATE}-{ITEM_KEY}-{USER_KEY}-{TIMEHASH}
+    */
+    const keyForUser = KeyUtils.getIdxKey(ENTITY.SAVED_POST, timeHash, userKey);
```
Comment (hewonjeong) [inline src/server/users/models.js]: key : savedPost-{userKey}-{itemKey}

File: src/server/items/controllers.js [hunk 4]
```diff
@@ -201,7 +201,7 @@ export default {
     const imageKey = `${ENTITY.IMAGE}-${timeHash}`;
     const imageIdxKey = KeyUtils.getIdxKey(ENTITY.IMAGE, timeHash, key);
     const item = {
-      key: key,
+      key,
```
Comment (ochanje210) [inline src/server/items/controllers.js]: why did you change it this way? key: key seems having reasonable readability

File: src/server/users/controllers.js [hunk 5]
```diff
@@ -2,7 +2,7 @@ import { KeyUtils, ENTITY, STATE } from '../key-utils';
 import { S3Connector } from '../aws-s3';
 import db from '../database';
 import { APIError } from '../ErrorHandler';
-import { CreatedPostManager, SavedPostManager } from './models';
+import { SavedPostManager, CreatedPostManager } from './models';
```
Comment (ochanje210) [inline src/server/users/controllers.js]: any specific reason for doing this?

File: src/server/items/controllers.js (hunk 4 above)
Comment (hewonjeong) [reply src/server/items/controllers.js]: To follow Airbnb Javascript style guide~~
https://github.com/airbnb/javascript#es6-object-concise

File: src/server/images/models.js [hunk 6]
```diff
@@ -23,4 +25,36 @@ export default class ImageManager {
...
+            keys.push(value.key);
+          });
+          resolve();
+        });
+      }));
+    }
+    Promise.all(promises)
+    .then(() => {
+      const s3Connector = new S3Connector();
+      const urls = (isThumbnail) ?
```
Comment (ochanje210) [inline src/server/images/models.js:47]: maybe you can add another condition and gives resized image to the frontend here?

File: src/server/items/controllers.js (hunk 4 above)
Comment (ochanje210) [reply src/server/items/controllers.js]: Thank you!

File: src/server/images/models.js (hunk 6 above)
Comment (hewonjeong) [reply src/server/images/models.js:47]: right. after then I think making other function for this then() block would be good.

File: src/server/users/controllers.js (hunk 5 above)
Comment (hewonjeong) [reply src/server/users/controllers.js]: When I removed 'addCreatedPost' function in controller, I needed to remove CreatedPostManager import.
After then I added CreatedPostManager import again to implement getCreatedPost function.
Not that big reason lol

File: src/server/users/models.js [hunk 7]
```diff
@@ -30,42 +33,153 @@ export default class UserManager {
     }
     return false;
   }
-  static getPostKeys(postType, userKey, cb) {
-    const prefix = `${postType}-${STATE.ALIVE}-${userKey}`;
-    fetchPrefix(prefix, (err, keyData) => {
-      cb(err, keyData);
+  static getPostKeys(postType, userKey) {
+    return new Promise((resolve, reject)=>{
```
Comment (ochanje210) [inline src/server/users/models.js:135]: Could you explain why this returns promise now?

File: src/test/node/items/controllers.js [hunk 8]
```diff
@@ -256,22 +257,34 @@ test('add an item to database', t => {
...
-        addedIdxItems.push(data.value);
-        if (data.value.key !== key) {
-          t.fail(`Indexing item's key is wrong : ${data.value.key}`);
-          t.end();
+      switch (entity) {
+      case ENTITY.ITEM:
+        if (isOriginKey) {
+          addedItem = data.value;
+          t.equal(data.value.address, expected.address, 'should be same address');
+          break;
```
Comment (ochanje210) [inline src/test/node/items/controllers.js]: This break is not necessary.

File: src/test/node/items/controllers.js [hunk 9]
```diff
@@ -256,22 +257,34 @@ test('add an item to database', t => {
...
+          if (data.value.key !== key) {
+            t.fail(`Indexing item's key is wrong : ${data.value.key}`);
+            t.end();
+          }
+        }
+        break;
+      case ENTITY.IMAGE:
+        if (isOriginKey) {
+          addedImage = data.value;
+          break;
```
Comment (ochanje210) [inline src/test/node/items/controllers.js]: ditto

File: src/test/node/users/controllers.js [hunk 10]
```diff
@@ -92,33 +94,34 @@ test('add a user to database', t => {
...
+      });
     });
   }).then(()=>{
     return new Promise((resolve, reject) => {
       const status = res.statusCode;
       const key = res._getData().data;
       t.equal(status, expected.status, 'should be same status');
       if (typeof key !== 'string' || !key.includes(testBody.userKey)) {
-        t.fail(`wrong key : ${key}`);
+        t.fail(`key is wrong: ${key}`);
```
Comment (ochanje210) [inline src/test/node/users/controllers.js]: wrong key sounds better?

File: src/server/images/models.js (hunk 6 above)
Comment (ochanje210) [reply src/server/images/models.js:47]: Awesome. Looking forward it

File: src/server/images/models.js (hunk 1 above)
Comment (ochanje210) [reply src/server/images/models.js]: Are you reviewing yourself?

File: src/server/users/controllers.js (hunk 5 above)
Comment (ochanje210) [reply src/server/users/controllers.js]: Ok I just wondered. I think it is better not to change it unless it has meanings

Comment (hewonjeong) [reply src/server/users/controllers.js]: Agree!

File: src/server/users/models.js (hunk 7 above)
Comment (hewonjeong) [reply src/server/users/models.js:135]: This function is used in savedPostManager and createdPostManager's getPosts function.
These functions are ran asynchronously.
If this function's return value is not Promise object, We have to use many callback and illegible code

File: src/test/node/items/controllers.js (hunk 8 above)
Comment (hewonjeong) [reply src/test/node/items/controllers.js]: good! thx!

File: src/test/node/users/controllers.js (hunk 10 above)
Comment (hewonjeong) [reply src/test/node/users/controllers.js]: hmm previous one is better?
I thought it was hard to understand if we got undefined value.
`wrong key : undefined`
it doesnt sound wired?

File: src/server/users/controllers.js (hunk 5 above)
Comment (ochanje210) [reply src/server/users/controllers.js]: I reverted back!

File: src/server/users/controllers.js [hunk 11]
```diff
@@ -2,7 +2,7 @@ import { KeyUtils, ENTITY, STATE } from '../key-utils';
 import { S3Connector } from '../aws-s3';
 import db from '../database';
 import { APIError } from '../ErrorHandler';
-import { SavedPostManager, CreatedPostManager } from './models';
+import { CreatedPostManager, SavedPostManager } from './models';
```
Comment (hewonjeong) [inline src/server/users/controllers.js]: thx~!

File: src/server/users/router.js [hunk 12]
```diff
@@ -97,39 +96,116 @@ router.post('/', controller.post);
...
+ *     HTTP/1.1 500 InternalError
  *     {
- *       error: "error message ..."
+ *       error: "database error"
  *     }
  */
-router.post('/savedpost', controller.addSavedPost);
-
-
+router.delete('/createdposts/', controller.deleteSavedPost);
```
Comment (ovekyc) [inline src/server/users/router.js]: '/createdposts'

File: src/test/node/users/controllers.js [hunk 13]
```diff
@@ -134,44 +137,123 @@ test('add created post using user controller', t => {
...
+    userKey: testUser.key,
+    // get a number of savedposts of test user
+    length: mockSavedPosts.filter((post)=>{
+      const prefix = `${ENTITY.SAVED_POST}-${STATE.ALIVE}-${testUser.key}`;
+      return (post.key.startsWith(prefix));
+    }).length
   };
+  const req = httpMocks.createRequest({
+    method: 'GET',
+    url: `/users/savedposts/${testUser.key}`,
```
Comment (ovekyc) [inline src/test/node/users/controllers.js]: /api/users/...

File: src/test/node/users/controllers.js [hunk 14]
```diff
@@ -134,44 +137,123 @@ test('add created post using user controller', t => {
...
+      STATE_STRING[STATE.ALIVE],
+      STATE_STRING[STATE.EXPIRED]
+    ]
   };
   const req = httpMocks.createRequest({
-    method: 'POST',
-    url: '/users/savedpost',
-    body: testBody
+    method: 'GET',
+    url: `/users/createdPosts/${testUser.key}`,
```
Comment (ovekyc) [inline src/test/node/users/controllers.js]: /api/...

File: src/test/node/users/controllers.js [hunk 15]
```diff
@@ -134,44 +137,123 @@ test('add created post using user controller', t => {
...
+    posts.map((post) => {
+      if (!post.imageUrl) {
+        t.fail('there is no imageUrl Field');
+      }
+      if (expected.states.indexOf(post.state) === -1) {
+        t.fail(`invalid state : ${post.state}`);
+      }
+      if (post.userKey !== expected.userKey) {
+        t.fail(`wrong user key : ${post.userKey}`);
+        return;
```
Comment (ovekyc) [inline src/test/node/users/controllers.js]: Why only return here?

File: src/test/node/users/controllers.js [hunk 16]
```diff
@@ -134,44 +137,123 @@ test('add created post using user controller', t => {
...
+test('delete saved post using user controller', t => {
+  const testUser = mockUsers[0].value;
+  const testItem = mockItems[0].value;
+  const expected = {
+    status: 200,
+    length: mockSavedPosts.length - 1
+  };
+  const req = httpMocks.createRequest({
+    method: 'DELETE',
+    url: '/users/savedposts',
```
Comment (ovekyc) [inline src/test/node/users/controllers.js]: /api/...

File: src/test/node/users/models.js [hunk 17]
```diff
@@ -26,45 +27,49 @@ const mockUser = {
   password: 'secret-password',
   imageUrl: 'url-of-image'
 };
-test('add created posts with user key', t => {
+test('add created post(CreatedPostManager.addPost', t => {
```
Comment (ovekyc) [inline src/test/node/users/models.js]: You maybe miss ')'

**Comment by ovekyc** [comment]
LGTM

**Comment by festinalente91** [comment]
lgtm
